### ISSUE TITLE
Work on a common coding style guide

### ISSUE THREAD

**Issue by purbon** [elastic/logstash#3462]
As big as the logstash code base is getting, and specially with having many contributors, lots of different coding styles will encounter. In order to simplify the codebase manageability and in order to simplify reviews we should work out a common coding style.

This style guide should help reviews and also be enforced throw tools like rubocop.

/cc @suyograo @jordansissel

**Comment by suyograo** [comment]
+1, how about using an existing style guide as a template and adding anything specific to LS? Like https://github.com/airbnb/ruby

**Comment by purbon** [comment]
@suyograo makes sense, there are several good ones, I will work on doing a first draft and then on top of that we can start discussing our specific LS customisations. Makes sense? I will also create a rubocop module so this could be enforced regularly. 

this would be so helpful for all of us :+1: :dancers: :+1:

**Comment by jordansissel** [comment]
I have a rubocop config in other projects, if you want a starting point.

**Comment by colinsurprenant** [comment]
I like this [guide format](https://github.com/bbatsov/ruby-style-guide) as it includes lots of good/bad examples and is pretty extensive. I did not read all rules in details but from what I saw it looked pretty decent. Maybe we could for that and adjust it to our own taste?

**Comment by jordansissel** [comment]
I may be misinformed, but I am pretty sure rubocop is just a codification
of that ruby community style guide. I will link in one of my rubocop
configs when I am home to show you where my preferences deviate(open for
discussion, also)

**Comment by purbon** [comment]
@colinsurprenant this is actually the other style guide I was talking about, what I will do is compose a first draft, including a rubocop conf that match it, and then throw a PR we can start having a discussion and finally get it into logstash. Makes sense?

**Comment by jordansissel** [comment]
Here's the rubocop config I've been using on other libraries: https://github.com/jordansissel/ruby-flores/blob/master/.rubocop.yml

Upon review, the above rubocop does me well except that I recently changed views on implicit returns, so I'd enable the RedundantReturn check instead of disabling it.

**Comment by jordansissel** [comment]
(for full disclosure, I am OK if we do not use anything from my rubocop config.)

**Comment by purbon** [comment]
@jordansissel thanks a lot for sharing, is much appreciate it. As I said, next step for me is to build a PR to update the style guide with input from all of this awesome resources plus a rubocop file, then we can start having a discussion to agree/include custom LS additions. Makes sense?

**Comment by jordansissel** [comment]
Yeah next steps:
* rubocop.yml PR
* figure out where we want to extend rubocop if needed to match out
style/practice recommendations

**Comment by purbon** [comment]
@jordansissel give me a minute and the initial draft PR is there :-)
